tests/benchmarks/wait_queues/src/main.c

```diff
@@ -0,0 +1,287 @@
+/*
+ * Copyright (c) 2024 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * @file
+ * This file contains tests that will measure the length of time required
+ * to add and remove threads from a wait queue that holds a varying number of
+ * threads. Each thread added to (and removed from) the wait queue is a dummy
+ * thread. As these dummy threads are inherently non-executable, this helps
+ * prevent the addition/removal of threads to/from the ready queue from being
+ * included in these measurements. Furthermore, the use of dummy threads helps
+ * reduce the memory footprint as not only are thread stacks not required,
+ * but we also do not need the full k_thread structure for each of these
+ * dummy threads.
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/timestamp.h>
+#include <zephyr/timing/timing.h>
+#include "utils.h"
+#include <zephyr/tc_util.h>
+#include <wait_q.h>
+#include <ksched.h>
+#include <stdio.h>
+
+uint32_t tm_off;
+
+static struct _thread_base dummy_thread[CONFIG_BENCHMARK_NUM_THREADS];
+static _wait_q_t wait_q;
+
+uint64_t add_cycles[CONFIG_BENCHMARK_NUM_THREADS];
+uint64_t remove_cycles[CONFIG_BENCHMARK_NUM_THREADS];
+
+/**
+ * Initialize each dummy thread.
+ */
+static void dummy_threads_init(unsigned int num_threads)
+{
+	unsigned int i;
+	unsigned int bucket_size;
+
+	bucket_size = (num_threads / CONFIG_NUM_PREEMPT_PRIORITIES) + 1;
+
+	for (i = 0; i < num_threads; i++) {
+		z_init_thread_base(&dummy_thread[i], i / bucket_size,
+				   _THREAD_DUMMY, 0);
+	}
+}
+
+static void cycles_reset(unsigned int num_threads)
+{
+	unsigned int i;
+
+	for (i = 0; i < num_threads; i++) {
+		add_cycles[i] = 0ULL;
+		remove_cycles[i] = 0ULL;
+	}
+}
+
+/**
+ * Each successive dummy thread added to the wait queue is either of the
+ * same or lower priority. Each dummy thread removed from the wait queue
+ * is of the same or lower priority than the one previous.
+ */
+static void test_decreasing_priority(_wait_q_t *q, unsigned int num_threads)
+{
+	unsigned int i;
+	timing_t start;
+	timing_t finish;
+
+	/* Add to tail of wait queue */
+
+	for (i = 0; i < num_threads; i++) {
+		start = timing_counter_get();
+		z_pend_thread((struct k_thread *)&dummy_thread[i],
+			      q, K_FOREVER);
+		finish = timing_counter_get();
+
+		add_cycles[i] += timing_cycles_get(&start, &finish);
+	}
+
+	/* Remove from head of wait queue */
+
+	for (i = 0; i < num_threads; i++) {
+		start = timing_counter_get();
+		z_unpend_thread((struct k_thread *)&dummy_thread[i]);
```

Comment:
Is this pretty common, or wouldn't getting the next thread to run be far more common than unpending?

Seems like we should test pending a bunch of threads, then running through getting them off the queue. Unless I've misunderstood the intent of z_unpend_thread()

Reply:
Numerically, the unpend case is called in a lot more contexts (basically any time something happens).  In practice, most threads never get unpended except when they get to run, obviously, so your intuition is probably correct.  But that operation is a rather more complicated scheduler path and not "wait_q" behavior appropriate for a microbenchmark (it's covered by other benchmarks basically).